### PR TITLE
Simplify error reporting in M4DApplication status

### PR DESCRIPTION
**Pull request by shlomitk1**
Fixes https://github.com/IBM/the-mesh-for-data/issues/67
This PR revisits error management by M4DApplication controller:
- Error condition reports non-fatal errors, such as grpc communication problems.
- Error condition causes a retrial.
- Failure condition reports fatal errors. No retrials are made.
- When applicable, an unified error report consisting of a dataset identifier, service/component from where the error came, a message itself.

### CODE REVIEW

File: manager/controllers/app/m4dapplication_controller_test.go [hunk 1]
```diff
@@ -228,22 +228,22 @@ var _ = Describe("M4DApplication Controller", func() {
...
+			Eventually(func() string {
 				f := &apiv1alpha1.M4DApplication{}
 				_ = k8sClient.Get(context.Background(), appSignature, f)
-				return f.Status.Conditions
+				return f.Status.Error
 			}, timeout, interval).ShouldNot(BeEmpty())
 
 			_ = k8sClient.Get(context.Background(), appSignature, resource)
-			Expect(resource.Status.Conditions[0].Reason).To(Equal("AccessDenied"))
+			Expect(resource.Status.Error).To(ContainSubstring("Governance policies forbid access to the data"))
```
Comment (roee88) [inline manager/controllers/app/m4dapplication_controller_test.go:239]: Too easy to break. Use consts for the error messages.

File: manager/apis/app/v1alpha1/m4dapplication_types.go [hunk 2]
```diff
@@ -61,34 +60,10 @@ type M4DApplicationSpec struct {
...
-	// Status of the condition: true or false
-	Status corev1.ConditionStatus `json:"status"`
-	// Reason is a short explanation of the reason for the current condition
-	// +optional
-	Reason string `json:"reason,omitempty"`
-	// Message contains the details of the current condition
+	// MaxRetries limits the number of failed attempts to generate a blueprint
 	// +optional
-	Message string `json:"message,omitempty"`
+	MaxRetries int64 `json:"maxRetries,omitempty"`
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/m4dapplication_types.go]: Removed max retries, returning error in reconcile instead.

File: manager/controllers/app/m4dapplication_controller.go [hunk 3]
```diff
@@ -297,20 +295,54 @@ func (r *M4DApplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 // GetAllModules returns all CRDs of the kind M4DModule mapped by their name
-func (r *M4DApplicationReconciler) GetAllModules() (map[string]*app.M4DModule, error) {
+func (r *M4DApplicationReconciler) GetAllModules() map[string]*app.M4DModule {
 	ctx := context.Background()
 
 	moduleMap := make(map[string]*app.M4DModule)
 	var moduleList app.M4DModuleList
 	if err := r.List(ctx, &moduleList); err != nil {
-		return moduleMap, err
+		r.Log.V(0).Info("Error while listing modules: " + err.Error())
```
[the remark this answers is not in the document]
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller.go:306]: Added back.

File: manager/controllers/app/m4dapplication_controller_test.go (hunk 1 above)
Comment (shlomitk1) [reply manager/controllers/app/m4dapplication_controller_test.go]: OK.

**Review by roee88 — APPROVED** [review]
- Try to avoid exposing internals to users (e.g. PolicyCompiler). 
- If you encounter too frequent retries then we can change the base in the exponential backoff to a higher value like 1s.
- Consider sticking to Error condition for any error reporting and use the other condition just as a boolean indicator (e.g.: Failed to explain that we won't retry or a positive field like Active if it makes sense)
- Just as info, once in every 24h the cache is invalidated and reconcile will be triggered anyway, which is a good thing because there are no real fatal errors, just those that require some human work.

Approving as is

**Comment by shlomitk1** [comment]
> 
> 
>     * Try to avoid exposing internals to users (e.g. PolicyCompiler).
> 
>     * If you encounter too frequent retries then we can change the base in the exponential backoff to a higher value like 1s.
> 
>     * Consider sticking to Error condition for any error reporting and use the other condition just as a boolean indicator (e.g.: Failed to explain that we won't retry or a positive field like Active if it makes sense)
> 
>     * Just as info, once in every 24h the cache is invalidated and reconcile will be triggered anyway, which is a good thing because there are no real fatal errors, just those that require some human work.
> 
> 
> Approving as is

Will be resolved in https://github.com/IBM/the-mesh-for-data/issues/153

**Pull request merged**